### DISCUSSION TITLE
Custom PMD rules for HTML

### DISCUSSION THREAD

**Discussion by DivyaSaravanan** [pmd/pmd#4657] [Q&A] [answered]
Is it possible to write custom PMD rules for HTML using Java?

**Comment by adangel** [comment, marked as answer]
Yes, this is possible.
See https://github.com/pmd/pmd/blob/master/pmd-html/src/main/java/net/sourceforge/pmd/lang/html/rule/bestpractices/UnnecessaryTypeAttributeRule.java for an example. The rule doc for this example is here: https://docs.pmd-code.org/latest/pmd_rules_html_bestpractices.html#unnecessarytypeattribute

**Comment by DivyaSaravanan** [comment]
What's the latest maven library that contains this change?

<dependency>
      <groupId>net.sourceforge.pmd</groupId>
      <artifactId>pmd-core</artifactId>
      <version>6.55.0</version>
 </dependency>

**Reply by adangel** [reply]
The latest version is currently 7.0.0-rc3

https://search.maven.org/search?q=g:net.sourceforge.pmd

You'll need pmd-core and pmd-html

**Comment by DivyaSaravanan** [comment]
Sample Rule:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="Sample PMD"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>CustomPMDRule</description>
    <rule name="CustomPMDRule"
          language="html"
          class="pmd.CustomPMDRule"
          message="Hello">
        <description>Custom HTML rule example</description>
    </rule>

</ruleset>
```

Java File:

```java
package pmd;


import net.sourceforge.pmd.lang.html.ast.ASTHtmlElement;
import net.sourceforge.pmd.lang.html.rule.AbstractHtmlRule;

public class CustomPMDRule extends AbstractHtmlRule {
    @Override
    public Object visit(ASTHtmlElement node, Object data) {
        System.out.println("Hello World");
        addViolation("Error in  HTML.", node);
        return super.visit(node, data);
    }

}
```

Execution :
`pmd check -d pathToSrcFile -R pathToxml`

Error:

Cannot load ruleset- An XML validation error occurred

Specifically it says :  Error at  **class="pmd.CustomPMDRule"**

Steps I followed to [install pmd](https://pmd.github.io/)

Is there anything I'm missing?

**Reply by DivyaSaravanan** [reply]
@adangel  could you please help?

**Reply by adangel** [reply]
Please have a look at this example:https://github.com/pmd/pmd-examples/blob/main/custom-rules/plain-java/README.md

I'v just updated it for the current version (7.0.0-rc3)

**Reply by DivyaSaravanan** [reply]
Thanks for sharing this. Can you please share the updated documentation using maven?

**Reply by DivyaSaravanan** [reply]
@adangel request your help here

**Reply by adangel** [reply]
https://github.com/pmd/pmd-examples/blob/main/custom-rules/maven-java/README.md

**Reply by adangel** [reply]
Hi @DivyaSaravanan , did it work for you meanwhile? Or is there another problem?
